### PR TITLE
Make all optimizers consistent so that they don't change gradients inplace

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30338 remove easy .data in torch/nn
* #30337 remote .dat from docs in torch/nn
* #30336 remove .data for torch/jit
* #30335 remove use of .data in torch/autograd
* #30334 remove dead python code that uses .data
* #30258 Remove use of .data in optimizers
* **#30257 Make all optimizers consistent so that they don't change gradients inplace**

Differential Revision: [D18665461](https://our.internmc.facebook.com/intern/diff/D18665461)

### CODE REVIEW

**Review by vincentqb — APPROVED** [review]
LGTM

**Review by anjali411 — APPROVED** [review]

**Comment by vincentqb** [comment]
As we discussed offline, since the gradient used to be modified in place, the gradient might have been different after running the `optimizer.step`. This is no longer the case.
